corehq/apps/hqcase/api.py

```diff
@@ -0,0 +1,210 @@
+
+from corehq.apps.hqcase.utils import submit_case_blocks
+from corehq.form_processor.interfaces.dbaccessors import CaseAccessors
+
+
+def serialize_case(case):
+    """Serializes a case for the V0.6 Case API"""
+    return {
+        "domain": case.domain,
+        "@case_id": case.case_id,
+        "@case_type": case.type,
+        "case_name": case.name,
+        "external_id": case.external_id,
+        "@owner_id": case.owner_id,
+        "date_opened": _isoformat(case.opened_on),
+        "last_modified": _isoformat(case.modified_on),
+        "server_last_modified": _isoformat(case.server_modified_on),
+        "closed": case.closed,
+        "date_closed": _isoformat(case.closed_on),
+        "properties": dict(case.dynamic_case_properties()),
+        "indices": {
+            index.identifier: {
+                "case_id": index.referenced_id,
+                "@case_type": index.referenced_type,
+                "@relationship": index.relationship,
+            }
+            for index in case.indices
+        }
+    }
+
+
+def _isoformat(value):
+    return value.isoformat() if value else None
+
+
+def is_simple_dict(d):
+    if not isinstance(d, dict) or not all(isinstance(v, str) for v in d.values()):
+        raise BadValueError("Case properties must be strings")
+
+
+class JsonIndex(jsonobject.JsonObject):
+    case_id = jsonobject.StringProperty()
+    temporary_id = jsonobject.StringProperty()
+    case_type = jsonobject.StringProperty(name='@case_type', required=True)
+    relationship = jsonobject.StringProperty(name='@relationship', required=True,
+                                             choices=('child', 'extension'))
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if not (bool(self.case_id) ^ bool(self.temporary_id)):
+            raise BadValueError("You must set either case_id or temporary_id, and not both.")
+
+
+class BaseJsonCaseChange(jsonobject.JsonObject):
+    case_name = jsonobject.StringProperty()
+    external_id = jsonobject.StringProperty()
+    user_id = jsonobject.StringProperty(required=True)
+    owner_id = jsonobject.StringProperty(name='@owner_id')
+    properties = jsonobject.DictProperty(validators=[is_simple_dict], default={})
+    indices = jsonobject.DictProperty(JsonIndex)
+    _is_case_creation = False
+
+    _allow_dynamic_properties = False
+
+    class Meta(object):
+        # prevent JsonObject from auto-converting dates etc.
+        string_conversions = ()
+
+    @memoized
+    def get_case_id(self):
+        return str(uuid.uuid4()) if self._is_case_creation else self.case_id
+
+    def get_caseblock(self):
+
+        def _if_specified(value):
+            return value if value is not None else CaseBlock.undefined
+
+        return CaseBlock(
+            case_id=self.get_case_id(),
+            user_id=self.user_id,
+            case_type=self.case_type if self._is_case_creation else CaseBlock.undefined,
```

Comment:
Is this check so that you don't inadvertently update the case type during an update?

Reply:
exactly

Reply:
Is that even a legal thing you can do with a form submission? 
If it is, then we should allow it here, right? 
If it isn't, then I suppose there is some error that is raised somewhere? I think we should catch this error and display it to the user, instead of silently modifying their submission.

Reply:
I thought we consider that unchangeable in HQ, and that was my intention here as well.  `case_type` is only defined in `JsonCaseCreation` (and dynamic properties are disabled), so attempting to include it in an update will fail.

Looks like it _is_ actually [allowed via form submission](https://github.com/dimagi/commcare-core/wiki/casexml20#case-action-elements).  I also see that save-to-case enables it, so I guess that settles the question.  I'll add that capacity in.